File: MenuFabrik/Services/MenuGeneratorEngine.swift
import Foundation

/// Moteur de génération basé sur un système de Scoring (Pondération) continu.
/// Toujours 100% découplé de la couche de persistance.
struct MenuGeneratorEngine {
    
    /// Assigne des recettes à une liste de repas
    static func generateMenu(for meals: [Meal], availableRecipes: [Recipe], participants: [Participant]) {
        // La logique d'exclusion globale est supprimée, car on gère ça plat par plat désormais
        
        // 3. Trier les repas chronologiquement
        let sortedMeals = meals.sorted {
            if Calendar.current.isDate($0.date, inSameDayAs: $1.date) {
                return $0.type == .lunch && $1.type == .dinner
            }
            return $0.date < $1.date
        }
        
        var usedRecipeIDs: Set<UUID> = []
        var previousCategory: RecipeCategory? = nil
        var previousSideDish: String? = nil
        
        for meal in sortedMeals {
            guard meal.status == .planned else { continue }
            
            let expectedType: MealType = (meal.type == .lunch) ? .lunch : .dinner
            let isWeekend = Calendar.current.isDateInWeekend(meal.date)
            
            // Seulement si le meal n'a pas encore de recette (pour un futur cas où on garde des repas verrouillés)
            guard meal.recipe == nil else { continue }
            
            // 1. Extraire les noms d'allergènes à éviter UNIQUEMENT pour les personnes présentes
            let attendees = meal.attendees ?? participants // Fallback sur tous les participants si `attendees` n'est pas rempli
            let excludedAllergens = Set(attendees.flatMap { $0.allergies.map { $0.name.lowercased() } })
            
            // 2. Filtrer les recettes sûres pour CE repas
            let safeRecipes = availableRecipes.filter { recipe in
                let recipeAllergens = Set(recipe.allergens.map { $0.name.lowercased() })
                return recipeAllergens.isDisjoint(with: excludedAllergens)
            }
            
            // On retient les scores pour chaque recette
            var scoredCandidates: [(recipe: Recipe, score: Int)] = []
            
            for recipe in safeRecipes {
                var score = 0
                
                // --- PÉNALITÉS ROUDHIBITOIRES (ou presque) ---
                if recipe.mealType != .both && recipe.mealType != expectedType {
                    score -= 100
                }
                
                if !isWeekend && recipe.requiresFreeTime {
                    score -= 50
                }
                
                if usedRecipeIDs.contains(recipe.id) {
                    score -= 80  // Diversité: Ne pas remanger la même chose la même semaine
                }
                
                if let prevCat = previousCategory, prevCat == recipe.category {
                    score -= 40  // Diversité: Ne pas manger la même catégorie de suite
                }
                
                // --- BONUS ---
                if meal.type == .lunch && recipe.prepTime <= 30 {
                    score += 20 // Idéal pour le midi
                }
                
                if isWeekend && recipe.requiresFreeTime {
                    score += 15 // C'est le bon moment pour la faire !
                }
                
                scoredCandidates.append((recipe, score))
            }
            
            // Tri décroissant pour avoir le meilleur score en premier
            scoredCandidates.sort { $0.score > $1.score }
            
            // S'il ne reste rien, tant pis, on passe
            guard !scoredCandidates.isEmpty else {
                print("Alerte: Pas de recettes compatibles pour \(meal.date)")
                continue
            }
            
            // On peut prendre aléatoirement parmi le Top 3 pour garder un peu de surprise
            let topScore = scoredCandidates[0].score
            // Les recettes dont le score est proche du top (marge de tolérance)
            let bestCandidates = scoredCandidates.filter { $0.score >= topScore - 10 }.map { $0.recipe }
            let chosenRecipe = bestCandidates.randomElement() ?? scoredCandidates[0].recipe
            
            meal.recipe = chosenRecipe
            
            // Amélioration: Tirage au sort de l'accompagnement (en évitant l'accompagnement précédent)
            var sides = chosenRecipe.suggestedSides
            if let prevSide = previousSideDish, sides.count > 1 {
                sides.removeAll { $0.name.lowercased() == prevSide.lowercased() }
            }
            if let randomSide = sides.randomElement() {
                meal.selectedSideDishes = [randomSide]
                previousSideDish = randomSide.name
            } else {
                meal.selectedSideDishes = []
                previousSideDish = nil
            }
        }
    }
    
    /// Trouve une recette alternative pour un repas unique
    static func generateAlternative(for meal: Meal, availableRecipes: [Recipe], participants: [Participant], history: [Meal]) {
        guard meal.status == .planned else { return }
        
        let attendees = meal.attendees ?? participants
        let excludedAllergens = Set(attendees.flatMap { $0.allergies.map { $0.name.lowercased() } })
        
        let safeRecipes = availableRecipes.filter { recipe in
            let recipeAllergens = Set(recipe.allergens.map { $0.name.lowercased() })
            return recipeAllergens.isDisjoint(with: excludedAllergens)
        }
        
        let expectedType: MealType = (meal.type == .lunch) ? .lunch : .dinner
        let isWeekend = Calendar.current.isDateInWeekend(meal.date)
        let usedRecipeIDs = Set(history.compactMap { $0.recipe?.id })
        let currentRecipeID = meal.recipe?.id
        
        var scoredCandidates: [(recipe: Recipe, score: Int)] = []
        
        for recipe in safeRecipes {
            var score = 0
            
            // Interdiction absolue de reproposer la recette qu'on vient de refuser
            if recipe.id == currentRecipeID {
                continue
            }
            
            if recipe.mealType != .both && recipe.mealType != expectedType {
                score -= 100
            }
            
            if !isWeekend && recipe.requiresFreeTime {
                score -= 50
            }
            
            if usedRecipeIDs.contains(recipe.id) {
                score -= 80
            }
            
            if meal.type == .lunch && recipe.prepTime <= 30 {
                score += 20
            }
            
            if isWeekend && recipe.requiresFreeTime {
                score += 15
            }
            
            scoredCandidates.append((recipe, score))
        }
        
        scoredCandidates.sort { $0.score > $1.score }
        
        guard !scoredCandidates.isEmpty else { return }
        
        let topScore = scoredCandidates[0].score
        let bestCandidates = scoredCandidates.filter { $0.score >= topScore - 10 }.map { $0.recipe }
        let chosenRecipe = bestCandidates.randomElement() ?? scoredCandidates[0].recipe
        
        meal.recipe = chosenRecipe
        if let randomSide = chosenRecipe.suggestedSides.randomElement() {
            meal.selectedSideDishes = [randomSide]
        } else {
            meal.selectedSideDishes = []
        }
    }
}

File: MenuFabrikWeb/src/services/MenuGeneratorEngine.ts
import { MealStatus, MealTime, type Meal } from '../models/Meal';
import { MealType, type Recipe, type RecipeCategory } from '../models/Recipe';
import type { Participant } from '../models/Participant';

export class MenuGeneratorEngine {

    /**
     * Assigne des recettes à une liste de repas
     */
    static generateMenu(meals: Meal[], availableRecipes: Recipe[], participants: Participant[]): void {
        const sortedMeals = [...meals].sort((a, b) => {
            const dateA = new Date(a.date);
            const dateB = new Date(b.date);

            if (dateA.toDateString() === dateB.toDateString()) {
                if (a.type === MealTime.LUNCH && b.type === MealTime.DINNER) return -1;
                if (a.type === MealTime.DINNER && b.type === MealTime.LUNCH) return 1;
                return 0;
            }
            return dateA.getTime() - dateB.getTime();
        });

        const usedRecipeIDs = new Set<string>();
        let previousCategories: RecipeCategory[] = [];
        let previousSideDishId: string | null = null;

        for (const meal of sortedMeals) {
            if (meal.format === 'note') continue;
            if (meal.status !== MealStatus.PLANNED) continue;
            if (meal.recipeId) continue;

            const safeRecipes = this.getSafeRecipes(meal, availableRecipes, participants);
            const scoredCandidates = this.scoreCandidates(
                meal,
                safeRecipes,
                usedRecipeIDs,
                previousCategories
            );

            if (scoredCandidates.length === 0) {
                console.warn(`Alerte: Pas de recettes compatibles pour le repas du ${meal.date}`);
                continue;
            }

            const chosenRecipe = this.pickBestCandidate(scoredCandidates);

            meal.recipeId = chosenRecipe.id;
            meal.recipe = chosenRecipe;

            if (chosenRecipe.id) {
                usedRecipeIDs.add(chosenRecipe.id);
            }
            previousCategories = chosenRecipe.categories;

            // Tirage au sort de l'accompagnement
            let sideIds = [...(chosenRecipe.suggestedSideIds || [])];
            if (previousSideDishId && sideIds.length > 1) {
                sideIds = sideIds.filter(id => id !== previousSideDishId);
            }

            if (sideIds.length > 0) {
                const randomSideId = sideIds[Math.floor(Math.random() * sideIds.length)];
                if (randomSideId) {
                    meal.selectedSideDishIds = [randomSideId];
                    previousSideDishId = randomSideId;
                } else {
                    meal.selectedSideDishIds = [];
                    previousSideDishId = null;
                }
            } else {
                meal.selectedSideDishIds = [];
                previousSideDishId = null;
            }
        }
    }

    /**
     * Trouve une recette alternative pour un repas unique
     */
    static generateAlternative(meal: Meal, availableRecipes: Recipe[], participants: Participant[], history: Meal[]): void {
        if (meal.format === 'note') return;
        if (meal.status !== MealStatus.PLANNED) return;

        const safeRecipes = this.getSafeRecipes(meal, availableRecipes, participants);

        const usedRecipeIDs = new Set<string>();
        history.forEach(h => {
            if (h.recipeId) usedRecipeIDs.add(h.recipeId);
        });

        const currentRecipeID = meal.recipeId;

        // Exclure la recette actuelle des candidats
        const candidates = safeRecipes.filter(r => r.id !== currentRecipeID);

        const scoredCandidates = this.scoreCandidates(
            meal,
            candidates,
            usedRecipeIDs,
            []
        );

        if (scoredCandidates.length === 0) return;

        const chosenRecipe = this.pickBestCandidate(scoredCandidates);

        meal.recipeId = chosenRecipe.id;
        meal.recipe = chosenRecipe;

        const sideIds = chosenRecipe.suggestedSideIds || [];
        if (sideIds.length > 0) {
            const randomSide = sideIds[Math.floor(Math.random() * sideIds.length)];
            if (randomSide) {
                meal.selectedSideDishIds = [randomSide];
            } else {
                meal.selectedSideDishIds = [];
            }
        } else {
            meal.selectedSideDishIds = [];
        }
    }

    // --- Private DRY Helpers ---

    private static getSafeRecipes(meal: Meal, availableRecipes: Recipe[], participants: Participant[]): Recipe[] {
        let attendees: Participant[] = [];
        if (meal.attendeeIds && meal.attendeeIds.length > 0) {
            attendees = participants.filter(p => meal.attendeeIds.includes(p.id!));
        } else {
            attendees = participants;
        }

        const excludedAllergenIds = new Set<string>();
        attendees.forEach(p => {
            p.allergyIds.forEach(id => excludedAllergenIds.add(id));
        });

        return availableRecipes.filter(recipe => {
            return !recipe.allergenIds.some(recipeAllergenId => excludedAllergenIds.has(recipeAllergenId));
        });
    }

    private static scoreCandidates(
        meal: Meal,
        recipes: Recipe[],
        usedRecipeIDs: Set<string>,
        previousCategories: RecipeCategory[]
    ): { recipe: Recipe, score: number }[] {
        const expectedType: MealType = (meal.type === MealTime.LUNCH) ? MealType.LUNCH : MealType.DINNER;
        const isWeekend = this.isWeekend(new Date(meal.date));

        const scoredCandidates: { recipe: Recipe, score: number }[] = [];

        for (const recipe of recipes) {
            let score = 0;

            if (recipe.mealType !== MealType.BOTH && recipe.mealType !== expectedType) {
                score -= 100;
            }

            if (!isWeekend && recipe.requiresFreeTime) {
                score -= 50;
            }

            if (recipe.id && usedRecipeIDs.has(recipe.id)) {
                score -= 80;
            }

            if (previousCategories.length > 0 && recipe.categories.some(c => previousCategories.includes(c))) {
                score -= 40;
            }

            if (meal.type === MealTime.LUNCH && recipe.prepTime <= 30) {
                score += 20;
            }

            if (isWeekend && recipe.requiresFreeTime) {
                score += 15;
            }

            scoredCandidates.push({ recipe, score });
        }

        return scoredCandidates.sort((a, b) => b.score - a.score);
    }

    private static pickBestCandidate(scoredCandidates: { recipe: Recipe, score: number }[]): Recipe {
        const topScore = scoredCandidates[0]!.score;
        const bestCandidates = scoredCandidates
            .filter(c => c.score >= topScore - 10)
            .map(c => c.recipe);

        return bestCandidates[Math.floor(Math.random() * bestCandidates.length)] || scoredCandidates[0]!.recipe;
    }

    private static isWeekend(date: Date): boolean {
        const day = date.getDay();
        return day === 0 || day === 6;
    }
}
